### ISSUE TITLE
staticGenerationBailout causes non-dismissable error overlay in dev mode

### ISSUE THREAD

**Issue by rijk** [vercel/next.js#43181]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
sandbox@sse-sandbox-fyyk70:/sandbox$ npx next info

    Operating System:
      Platform: linux
      Arch: x64
      Version: #45~20.04.1-Ubuntu SMP Mon Apr 4 09:38:31 UTC 2022
    Binaries:
      Node: 16.15.1
      npm: 8.11.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.0.5-canary.3
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Describe the Bug

When using `cookies()` anywhere inside a (potentially) static layout, next throws an `Error: Dynamic server usage: cookies`.

Example use case:

- /[organization]/homepage, /[organization]/about, … → static, `generateStaticParams` implemented in organization segment layout
- /[organization]/admin → fully dynamic, as data depends on the user who is logged in (through cookie)

### Expected Behavior

I would expect Next to just exit out of static generation for this `/[organization]/admin` subpath, without throwing an error.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://codesandbox.io/s/blissful-napier-fyyk70?file=/app/%5Borganization%5D/admin/page.tsx

### To Reproduce

Go to https://fyyk70.sse.codesandbox.io/myorg/admin

**Comment by rijk** [comment]
Possibly related to #43148

**Comment by lassegit** [comment]
I have the same issue caused by running a custom `getToken` function that uses the `cookies` from `next/headers` which gets called from the `RootLayout` (calling it from pages higher up doesn't cause an issue).

`generateStaticParams` is called from a `posts[slug]` sub directory. 

Version: 13.0.6.

**Comment by rijk** [comment]
Have you found a way to work around it? It is currently blocking using static rendering for me.

**Comment by lassegit** [comment]
Then having a "dynamic" header element that uses cookies will prevent this.

**Comment by rijk** [comment]
Actually, the same happens when using a server component anywhere in a static route. As @Arctomachine mentioned it only happens running `next dev`, when running `next build` and `next start` everything works fine, and the page is just not statically generated. I believe this throwing of an error by `staticGenerationBailout` is actually a feature, that allows Next to detect which pages not to render statically. I have a path where a server component is rendered conditionally, and only the pages on which it's rendered are _not_ static, which is quite cool.

I think the issue is just that this error accidentally crashes dev mode. cc @timneutkens 

```
client.js?234c:691 Uncaught Error: Dynamic server usage: headers
    at Object.staticGenerationBailout
```

**Comment by rijk** [comment]
This seems to be fixed in the 13.0.7, I can't reproduce it anymore locally or in the Sandbox.

**Comment by rijk** [comment]
Sorry, reopening as I can definitely still reproduce on 13.0.7 (locally and in Sandbox: https://fyyk70.sse.codesandbox.io/myorg/admin).

**Comment by timneutkens** [comment]
Created a new sandbox to reproduce it as yours is using the old sandbox format: https://codesandbox.io/p/sandbox/throbbing-butterfly-w1lbze?file=%2Fapp%2F%5Borganization%5D%2Fadmin%2Fpage.js&selection=%5B%7B%22endColumn%22%3A2%2C%22endLineNumber%22%3A9%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A7%7D%5D

The issue is that you're using `headers()` combined with `generateStaticParams` and it's not error out with a clear error.
Basically `headers()` is at odds with `generateStaticParams()` because you're basically saying: "this page is static" and then while rendering it's actually not static because `headers()` indicates "this page is dynamic".

**Comment by returnednull** [comment]
I am encountering with same issue. I'm using static generated locale paths on beginning on my route and I can't opt-in back to dynamic rendering no matter what. Dev overlay says: 
`Error: A required parameter (slug) was not provided as a string in generateStaticParams for /[lng]/destinations/[slug]`
`[lng]` my static path which is generated on build time.
`[slug]` is the route I want to opt in back dynamic rendering. I can continue with static rendering but there will frequently changing thousands of slug page will be generated which I dont want to do that.
I can supress error with return `[{ slug: '' }]` from `generateStaticParams` and it still uses dynamic fetching in this way. I upgraded from 13.0.6 to 13.1.1 and it didn't change anything. I'll continue with tricky way but I can share MRE sandbox if it will be helpfull. I've different fetching functions and static pages with different caching/revalidating configurations.

**Comment by rijk** [comment]
Yeah I've encountered that as well, when using ["top down" generation](https://beta.nextjs.org/docs/api-reference/generate-static-params#generate-segments-from-the-top-down). Note that you can also silence it by just `return []`. I've been wondering why this isn't the default behavior, but I think this is done deliberately to alert you a subsection of your site is not static.

By the way, to correct [my previous comment](https://github.com/vercel/next.js/issues/43181#issuecomment-1357552850): these segments are **not** generated dynamically on Vercel but fail with a 500 there as well. So I am not sure this issue is actually an issue. I'll close it for now.

**Comment by logemann** [comment]
So is this a bug or something?  This is my case... as soon as i add generateStaticParams and starting in dev mode, i get the error too because of createServerClient() which uses cookies.

```
const getProperty  = async (id: string) => {
   const supabase = createServerClient();
   const { data } = await supabase.from('properties').select('*').eq("id", id).single();
   return data;
};

export async function generateStaticParams() {
   return [{id: "1"}];
}

export default async function  PropertySinglePage({params}: {params: any}) {
const {id}: {id: string} = params;
   const entity = await getProperty(id);
...
}
```
